pkg/kratos/service.go

```diff
@@ -226,7 +256,153 @@ func (s *Service) CreateBrowserLoginFlow(
+	ctx, span := s.tracer.Start(ctx, "kratos.Service.CreateBrowserRegistrationFlow")
+	defer span.End()
+
+	request := s.kratos.FrontendApi().
+		CreateBrowserRegistrationFlow(ctx).
+		ReturnTo(returnTo)
+
+	flow, resp, err := s.kratos.FrontendApi().CreateBrowserRegistrationFlowExecute(request)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return flow, resp.Cookies(), nil
+}
+
+func (s *Service) UpdateRegistrationFlow(ctx context.Context, flowId string, body kClient.UpdateRegistrationFlowBody, cookies []*http.Cookie) (*RegistrationFlowResponse, []*http.Cookie, error) {
+	ctx, span := s.tracer.Start(ctx, "kratos.Service.UpdateRegistrationFlow")
+	defer span.End()
+
+	var registration *RegistrationFlowResponse = nil
+
+	request := s.kratos.FrontendApi().
+		UpdateRegistrationFlow(ctx).
+		Flow(flowId).
+		UpdateRegistrationFlowBody(body).
+		Cookie(httpHelpers.CookiesToString(cookies))
+
+	successfulFlow, resp, err := s.kratos.FrontendApi().UpdateRegistrationFlowExecute(request)
+
+	if resp != nil {
+		cookies = resp.Cookies()
+	}
+
+	if successfulFlow != nil {
+		registration = &RegistrationFlowResponse{completedFlow: successfulFlow}
+		return registration, cookies, err
+	}
+
+	// if we get an error from the sdk + one of the expected status codes, then the flow will be in the response body
+	if err != nil && resp != nil && slices.Contains([]int{http.StatusOK, http.StatusUnprocessableEntity, http.StatusBadRequest}, resp.StatusCode) {
+		registration, err = s.tryProcessingRegistration(resp)
+	}
+
+	return registration, cookies, err
+}
+
+func (s *Service) tryProcessingRegistration(resp *http.Response) (*RegistrationFlowResponse, error) {
+	responseBody, err := io.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error marshalling response body: %w", err)
+	}
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		var completedFlow kClient.SuccessfulNativeRegistration
+		if err := json.Unmarshal(responseBody, &completedFlow); err != nil {
+			return nil, fmt.Errorf("unexpected error: %w", err)
+		}
+
+		return &RegistrationFlowResponse{completedFlow: &completedFlow}, nil
+
+	case http.StatusBadRequest:
+		var registrationFlow kClient.RegistrationFlow
+		// in this case the error has a body with the actual flow to return or user is already authenticated
+		if err := json.Unmarshal(responseBody, &registrationFlow); err != nil {
+			if isSessionAlreadyAvailableError(responseBody) {
+				return nil, fmt.Errorf("a valid session was detected and thus registration is not possible")
+			}
+
+			return nil, fmt.Errorf("unexpected error unmarshalling response body: %w", err)
+		}
+
+		if isDuplicateIdentifierError(registrationFlow) {
+			return nil, fmt.Errorf("an account with the same identifier already exists, contact support")
+		}
+
+		if state, ok := registrationFlow.State.(string); ok && state == "choose_method" {
```

Comment:
thought: I thought we didn't want to map kratos error responses to login UI success responses. A comment here explaining the reason for this would be helpful